validator.go

```diff
@@ -0,0 +1,267 @@
+	vErr := new(ValidationError)
+
+	// Check, if we have a time func
+	if v.timeFunc != nil {
+		now = v.timeFunc()
+	} else {
+		now = time.Now()
+	}
+
+	// We always need to check the expiration time, but the claim itself is OPTIONAL
+	if !v.VerifyExpiresAt(claims, now, false) {
+		vErr.Inner = ErrTokenExpired
+		vErr.Errors |= ValidationErrorExpired
+	}
+
+	// We always need to check not-before, but the claim itself is OPTIONAL
+	if !v.VerifyNotBefore(claims, now, false) {
+		vErr.Inner = ErrTokenNotValidYet
+		vErr.Errors |= ValidationErrorNotValidYet
+	}
+
+	// Check issued-at if the option is enabled
+	if v.verifyIat && !v.VerifyIssuedAt(claims, now, false) {
+		vErr.Inner = ErrTokenUsedBeforeIssued
+		vErr.Errors |= ValidationErrorIssuedAt
+	}
+
+	// If we have an expected audience, we also require the audience claim
+	if v.expectedAud != "" && !v.VerifyAudience(claims, v.expectedAud, true) {
+		vErr.Inner = ErrTokenInvalidAudience
+		vErr.Errors |= ValidationErrorAudience
+	}
+
+	// If we have an expected issuer, we also require the issuer claim
+	if v.expectedIss != "" && !v.VerifyIssuer(claims, v.expectedIss, true) {
+		vErr.Inner = ErrTokenInvalidIssuer
+		vErr.Errors |= ValidationErrorIssuer
+	}
+
+	// If we have an expected subject, we also require the subject claim
+	if v.expectedSub != "" && !v.VerifySubject(claims, v.expectedSub, true) {
+		vErr.Inner = ErrTokenInvalidSubject
+		vErr.Errors |= ValidationErrorSubject
+	}
+
+	// Finally, we want to give the claim itself some possibility to do some
+	// additional custom validation based on their custom claims
+	cvt, ok := claims.(CustomClaims)
+	if ok {
+		if err := cvt.CustomValidation(); err != nil {
+			vErr.Inner = err
+			vErr.Errors |= ValidationErrorClaimsInvalid
+		}
+	}
+
+	if vErr.valid() {
+		return nil
+	}
+
+	return vErr
+}
+
+// VerifyAudience compares the aud claim against cmp.
+// If required is false, this method will return true if the value matches or is unset
+func (v *Validator) VerifyAudience(claims Claims, cmp string, req bool) bool {
+	aud, err := claims.GetAudience()
+	if err != nil {
+		return false
+	}
+
+	return verifyAud(aud, cmp, req)
+}
+
+// VerifyExpiresAt compares the exp claim against cmp (cmp < exp).
+// If req is false, it will return true, if exp is unset.
+func (v *Validator) VerifyExpiresAt(claims Claims, cmp time.Time, req bool) bool {
+	var time *time.Time = nil
+
+	exp, err := claims.GetExpirationTime()
+	if err != nil {
+		return false
+	} else if exp != nil {
+		time = &exp.Time
+	}
+
+	return verifyExp(time, cmp, req, v.leeway)
+}
+
+// VerifyIssuedAt compares the iat claim against cmp (cmp >= iat).
+// If req is false, it will return true, if iat is unset.
+func (v *Validator) VerifyIssuedAt(claims Claims, cmp time.Time, req bool) bool {
+	var time *time.Time = nil
+
+	iat, err := claims.GetIssuedAt()
+	if err != nil {
+		return false
+	} else if iat != nil {
+		time = &iat.Time
+	}
+
+	return verifyIat(time, cmp, req, v.leeway)
```

Comment:
I could find the leeway and validation suggested only for `exp` and `nbf` registered claims in the [rfc](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6), but not for `iat`.

```
This claim can be used to determine the age of the JWT.
```

Just pointing here, in case we want to avoid verification if IssueAt claim

Reply:
That's a good point. 

I think we could argue either way. In the standard, this is just an information about the age and no check of any kind is required or mentioned. On the other hand, if we want to check for plausibility of that age it might make sense to include the same leeway as for the other time-related properties.

Reply:
It would make sense to me to make this validation optional. This would prevent any possible issue with Identity Providers (JWT issuers).

Reply:
It is. You need to specify the `WithIssuedAt` option in the validator, otherwise it will not be validated.

Reply:
> It would make sense to me to make this validation optional.

Sorry, my comment was not correct. I meant to say the leeway should be optional, not just the validation.

Reply:
Well it is optional, if you don't specify any leeway it will check for the strict time on any of the time-based claims. Only making it "even more optional" for IAT would be overkill in my opinion.